FWCore/Framework/src/SubProcess.cc

```diff
@@ -379,23 +383,30 @@ namespace edm {
     if (subProcesses_.empty()) {
       afterProcessTask = std::move(finalizeEventTask);
     } else {
-      afterProcessTask = WaitingTaskHolder(make_waiting_task(
-          tbb::task::allocate_root(), [this, &ep, finalizeEventTask](std::exception_ptr const* iPtr) mutable {
-            if (not iPtr) {
-              for (auto& subProcess : boost::adaptors::reverse(subProcesses_)) {
-                subProcess.doEventAsync(finalizeEventTask, ep);
-              }
-            } else {
-              finalizeEventTask.doneWaiting(*iPtr);
-            }
-          }));
+      afterProcessTask = WaitingTaskHolder(
+          make_waiting_task(tbb::task::allocate_root(),
+                            [this, &ep, finalizeEventTask, iEventSetupImpls](std::exception_ptr const* iPtr) mutable {
+                              if (not iPtr) {
+                                for (auto& subProcess : boost::adaptors::reverse(subProcesses_)) {
+                                  subProcess.doEventAsync(finalizeEventTask, ep, iEventSetupImpls);
+                                }
+                              } else {
+                                finalizeEventTask.doneWaiting(*iPtr);
+                              }
+                            }));
     }
 
-    schedule_->processOneEventAsync(
-        std::move(afterProcessTask), ep.streamID().value(), ep, esp_->eventSetup(), serviceToken_);
+    schedule_->processOneEventAsync(std::move(afterProcessTask),
+                                    ep.streamID().value(),
+                                    ep,
+                                    *iEventSetupImpls->at(esp_->subProcessIndex()),
```

Comment:
No reason to use `at`

Reply:
Why is the `EventSetupImpl` gotten from the list here but in other calls it is gotten from `esp_->eventSetupImpl()`?

Reply:
First, I will change all the calls to the function named 'at' to use square brackets like an array access in SubProcess.cc. They were just sanity checks that should not ever fail.

The second question is more difficult. Passing the EventSetupImpl objects around is handled one way in all the Run transitions and in a different way in all the luminosity block transitions. This is because multiple luminosity blocks can be processed concurrently, but at the present time runs can only be processed one at a time. The EventSetupProvider knows the most recently setup EventSetupImpl and that is all that is needed for Runs. It will not change until the processing for the run is complete. It is simpler to get the EventSetupImpl from there. Plus that was the we way were doing it before, so it minimizes the changes in this PR. For LuminosityBlocks this will not work because there can be multiple luminosity blocks in flight at the same time and they may be associated with different EventSetupImpl objects. The LuminosityBlockProcessingStatus objects are the things that hold the EventSetupImpl's in this new code. Currently RunProcessingStatus objects do not exist and the machinery associated with them does not exist yet either.

If and when we implement the code to support concurrent Runs, then this will need to change. We will likely implement RunProcessingStatus objects and all the associated machinery. Then we will also need to get the EventSetupImpl object in the same way for Runs as done in this PR for LuminosityBlocks. It seemed to me this was better left for a separate PR.

Reply:
OK